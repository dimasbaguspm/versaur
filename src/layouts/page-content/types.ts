import type { ReactNode, HTMLAttributes } from 'react'

/**
 * Props for PageContent main container
 *
 * A layout component that provides consistent horizontal spacing
 * matching the page-header component, with additional vertical
 * padding for content separation
 */
export interface PageContentProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * Content to be displayed inside the page content area
   */
  children?: ReactNode
}
