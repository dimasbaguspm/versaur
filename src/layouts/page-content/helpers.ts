import { cva } from 'class-variance-authority'

/**
 * PageContent root styles
 * Provides consistent horizontal spacing matching page-header
 * with additional vertical padding for content separation
 */
export const pageContentVariants = cva('w-full px-4 sm:px-6 py-6 sm:py-8')
