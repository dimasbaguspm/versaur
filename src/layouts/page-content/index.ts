export { PageContent } from './page-content'
export type { PageContentProps } from './types'
