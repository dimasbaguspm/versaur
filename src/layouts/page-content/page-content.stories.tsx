import type { Meta, StoryObj } from '@storybook/react'
import { PageContent } from './page-content'
import { Text } from '@/primitive/text'
import { NoResults } from '@/primitive/no-results'
import { Button } from '@/primitive/button'
import { TextInput } from '@/forms/text-input'
import { EmailInput } from '@/forms/email-input'
import { SearchInput } from '@/forms/search-input'
import { TextAreaInput } from '@/forms/textarea-input'
import { SearchIcon, UserPlus } from 'lucide-react'

/**
 * PageContent provides consistent horizontal spacing that matches the page-header
 * component, with additional vertical padding for proper content separation.
 * Perfect for creating cohesive page layouts when used alongside PageHeader.
 */
const meta = {
  title: 'Layouts/PageContent',
  component: PageContent,
  parameters: {
    layout: 'fullscreen',
    docs: {
      description: {
        component: `
PageContent is a layout component designed to be a sibling of PageHeader. It provides:
- Consistent horizontal spacing matching PageHeader (px-4 sm:px-6)
- Additional vertical padding for content separation (py-6 sm:py-8)
- Mobile-first responsive design
- Full width container for all page content

Use PageContent to wrap your main page content for consistent layouts.
        `,
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    children: {
      control: { type: 'text' },
      description: 'Content to be displayed inside the page content area',
    },
    className: {
      control: { type: 'text' },
      description: 'Additional CSS classes to apply to the component',
    },
  },
} satisfies Meta<typeof PageContent>

export default meta
type Story = StoryObj<typeof meta>

/**
 * Basic usage of PageContent with Versaur Text components.
 * Shows the default spacing and responsive behavior using semantic typography.
 */
export const Default: Story = {
  args: {
    children: (
      <div className='space-y-6'>
        <div className='prose'>
          <Text as='p' fontSize='base' color='ghost'>
            This is the main content area of your page. PageContent provides
            consistent horizontal spacing that matches the PageHeader component,
            ensuring a cohesive layout across your application.
          </Text>
        </div>
        <div className='grid gap-4 md:grid-cols-2 lg:grid-cols-3'>
          <div className='p-4 bg-neutral rounded-lg'>
            <Text as='h3' fontSize='lg' fontWeight='semibold' className='mb-2'>
              Card 1
            </Text>
            <Text as='p' color='gray'>
              Some content in the first card using Versaur Text component.
            </Text>
          </div>
          <div className='p-4 bg-neutral rounded-lg'>
            <Text as='h3' fontSize='lg' fontWeight='semibold' className='mb-2'>
              Card 2
            </Text>
            <Text as='p' color='gray'>
              Some content in the second card with semantic typography.
            </Text>
          </div>
          <div className='p-4 bg-neutral rounded-lg'>
            <Text as='h3' fontSize='lg' fontWeight='semibold' className='mb-2'>
              Card 3
            </Text>
            <Text as='p' color='gray'>
              Some content in the third card demonstrating consistency.
            </Text>
          </div>
        </div>
      </div>
    ),
  },
}

/**
 * PageContent with Versaur form components, showing how it works with
 * interactive components and maintains proper spacing.
 */
export const WithFormElements: Story = {
  args: {
    children: (
      <div className='space-y-8'>
        <div>
          <Text as='h2' fontSize='xl' fontWeight='semibold' className='mb-4'>
            User Settings
          </Text>
          <div className='space-y-6 max-w-lg'>
            <TextInput
              label='Full Name'
              placeholder='Enter your full name'
              variant='primary'
            />
            <EmailInput
              label='Email Address'
              placeholder='Enter your email'
              variant='primary'
            />
            <TextAreaInput
              label='Bio'
              placeholder='Tell us about yourself'
              variant='primary'
              minRows={4}
              maxRows={6}
            />
            <div className='flex gap-3'>
              <Button variant='primary' size='md'>
                Save Changes
              </Button>
              <Button variant='ghost' size='md'>
                Cancel
              </Button>
            </div>
          </div>
        </div>
      </div>
    ),
  },
}

/**
 * PageContent with NoResults component, showing how it handles
 * empty states with proper spacing and semantic structure.
 */
export const WithNoResults: Story = {
  args: {
    children: (
      <NoResults
        icon={SearchIcon}
        title='No Content Yet'
        subtitle='This area is waiting for content. Add some items to get started.'
        action={
          <Button variant='primary' size='md'>
            Get Started
          </Button>
        }
      />
    ),
  },
}

/**
 * PageContent with search and data display, demonstrating how it works
 * with complex layouts including search functionality.
 */
export const WithSearchAndData: Story = {
  args: {
    children: (
      <div className='space-y-6'>
        <div className='flex flex-col sm:flex-row gap-4 justify-between items-start sm:items-center'>
          <Text as='h2' fontSize='xl' fontWeight='semibold'>
            Users
          </Text>
          <div className='flex gap-3 w-full sm:w-auto'>
            <SearchInput
              placeholder='Search users...'
              className='flex-1 sm:w-64'
            />
            <Button variant='primary' size='md'>
              <UserPlus className='w-4 h-4 mr-2' />
              Add User
            </Button>
          </div>
        </div>
        <div className='overflow-x-auto'>
          <table className='w-full border-collapse'>
            <thead>
              <tr className='border-b border-border'>
                <th className='text-left py-3 px-4'>
                  <Text as='span' fontWeight='medium'>
                    Name
                  </Text>
                </th>
                <th className='text-left py-3 px-4'>
                  <Text as='span' fontWeight='medium'>
                    Email
                  </Text>
                </th>
                <th className='text-left py-3 px-4'>
                  <Text as='span' fontWeight='medium'>
                    Role
                  </Text>
                </th>
                <th className='text-left py-3 px-4'>
                  <Text as='span' fontWeight='medium'>
                    Status
                  </Text>
                </th>
                <th className='text-left py-3 px-4'>
                  <Text as='span' fontWeight='medium'>
                    Actions
                  </Text>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr className='border-b border-border'>
                <td className='py-3 px-4'>
                  <Text as='span'>John Doe</Text>
                </td>
                <td className='py-3 px-4'>
                  <Text as='span' color='gray'>
                    john@example.com
                  </Text>
                </td>
                <td className='py-3 px-4'>
                  <Text as='span'>Admin</Text>
                </td>
                <td className='py-3 px-4'>
                  <span className='px-2 py-1 bg-success-soft text-success-bold rounded text-sm'>
                    <Text as='span' fontSize='sm' color='success'>
                      Active
                    </Text>
                  </span>
                </td>
                <td className='py-3 px-4'>
                  <Button variant='ghost' size='sm'>
                    Edit
                  </Button>
                </td>
              </tr>
              <tr className='border-b border-border'>
                <td className='py-3 px-4'>
                  <Text as='span'>Jane Smith</Text>
                </td>
                <td className='py-3 px-4'>
                  <Text as='span' color='gray'>
                    jane@example.com
                  </Text>
                </td>
                <td className='py-3 px-4'>
                  <Text as='span'>Editor</Text>
                </td>
                <td className='py-3 px-4'>
                  <span className='px-2 py-1 bg-warning-soft text-warning-bold rounded text-sm'>
                    <Text as='span' fontSize='sm' color='warning'>
                      Pending
                    </Text>
                  </span>
                </td>
                <td className='py-3 px-4'>
                  <Button variant='ghost' size='sm'>
                    Edit
                  </Button>
                </td>
              </tr>
              <tr className='border-b border-border'>
                <td className='py-3 px-4'>
                  <Text as='span'>Mike Johnson</Text>
                </td>
                <td className='py-3 px-4'>
                  <Text as='span' color='gray'>
                    mike@example.com
                  </Text>
                </td>
                <td className='py-3 px-4'>
                  <Text as='span'>Viewer</Text>
                </td>
                <td className='py-3 px-4'>
                  <span className='px-2 py-1 bg-success-soft text-success-bold rounded text-sm'>
                    <Text as='span' fontSize='sm' color='success'>
                      Active
                    </Text>
                  </span>
                </td>
                <td className='py-3 px-4'>
                  <Button variant='ghost' size='sm'>
                    Edit
                  </Button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    ),
  },
}
