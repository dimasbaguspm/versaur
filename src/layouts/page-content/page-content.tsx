import { forwardRef } from 'react'
import { cn } from '@/utils/cn'
import { pageContentVariants } from './helpers'
import type { PageContentProps } from './types'

/**
 * PageContent - Layout component for page content areas
 *
 * Provides consistent horizontal spacing that matches the page-header
 * component, with additional vertical padding for proper content separation.
 * Designed to be used as a sibling to PageHeader for consistent page layouts.
 *
 * Key features:
 * - Horizontal padding matches page-header (px-4 sm:px-6)
 * - Additional vertical padding for content separation (py-6 sm:py-8)
 * - Mobile-first responsive design
 * - Full width container
 */
export const PageContent = forwardRef<HTMLDivElement, PageContentProps>(
  ({ children, className, ...props }, ref) => {
    return (
      <div
        ref={ref}
        className={cn(pageContentVariants(), className)}
        {...props}
      >
        {children}
      </div>
    )
  }
)
