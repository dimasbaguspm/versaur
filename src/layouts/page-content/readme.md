# PageContent

A layout component that provides consistent horizontal spacing matching the PageHeader component, with additional vertical padding for proper content separation.

## Usage

```tsx
import { PageContent } from '@/layouts/page-content'

function MyPage() {
  return (
    <>
      <PageHeader title="Dashboard" />
      <PageContent>
        <div className="space-y-6">
          <p>Your main page content goes here.</p>
          {/* More content */}
        </div>
      </PageContent>
    </>
  )
}
```

## Features

- **Consistent Spacing**: Horizontal padding matches PageHeader (px-4 sm:px-6)
- **Content Separation**: Additional vertical padding (py-6 sm:py-8) for proper content spacing
- **Mobile-First**: Responsive design that works across all screen sizes
- **Full Width**: Takes full width of its container
- **Accessible**: Semantic HTML structure

## Design System

PageContent is designed to be a sibling component to PageHeader, ensuring consistent horizontal alignment and proper vertical spacing throughout your application layout.

### Spacing Hierarchy

- **Horizontal**: Same as PageHeader (px-4 on mobile, px-6 on desktop)
- **Vertical**: More generous padding (py-6 on mobile, py-8 on desktop) for content breathing room

## Props

Extends all standard HTML div element props.

| Prop | Type | Description |
|------|------|-------------|
| `children` | `ReactNode` | Content to be displayed inside the page content area |
| `className` | `string` | Additional CSS classes to apply |

## Examples

### Basic Usage
Simple content wrapper with consistent spacing.

### With Form Elements
Works great with forms and interactive elements.

### Minimal Content
Handles sparse content gracefully with proper spacing.
