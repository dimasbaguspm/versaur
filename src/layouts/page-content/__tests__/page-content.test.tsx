import { render } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import { describe, expect, it } from 'vitest'
import * as Stories from '../page-content.stories'

const { Default, WithFormElements, WithNoResults, WithSearchAndData } =
  composeStories(Stories)

describe('PageContent', () => {
  it('renders default story correctly', () => {
    const { container } = render(<Default />)
    expect(container.firstChild).toHaveClass(
      'w-full',
      'px-4',
      'sm:px-6',
      'py-6',
      'sm:py-8'
    )
  })

  it('renders with form elements story correctly', () => {
    const { container } = render(<WithFormElements />)
    expect(container.firstChild).toHaveClass(
      'w-full',
      'px-4',
      'sm:px-6',
      'py-6',
      'sm:py-8'
    )
  })

  it('renders minimal story correctly', () => {
    const { container } = render(<WithNoResults />)
    expect(container.firstChild).toHaveClass(
      'w-full',
      'px-4',
      'sm:px-6',
      'py-6',
      'sm:py-8'
    )
  })

  it('renders search and data story correctly', () => {
    const { container } = render(<WithSearchAndData />)
    expect(container.firstChild).toHaveClass(
      'w-full',
      'px-4',
      'sm:px-6',
      'py-6',
      'sm:py-8'
    )
  })

  it('renders children correctly', () => {
    const { getByText } = render(<Default />)
    expect(getByText(/This is the main content area/)).toBeInTheDocument()
    expect(getByText('Card 1')).toBeInTheDocument()
  })

  it('renders NoResults story correctly', () => {
    const { getByText } = render(<WithNoResults />)
    expect(getByText('No Content Yet')).toBeInTheDocument()
    expect(getByText('Get Started')).toBeInTheDocument()
  })

  it('renders form elements story correctly', () => {
    const { getByText } = render(<WithFormElements />)
    expect(getByText('User Settings')).toBeInTheDocument()
    expect(getByText('Save Changes')).toBeInTheDocument()
  })

  it('applies custom className', () => {
    const customClass = 'custom-page-content'
    const { container } = render(<Default className={customClass} />)
    expect(container.firstChild).toHaveClass(customClass)
  })

  it('matches snapshot', () => {
    expect(render(<Default />).asFragment()).toMatchSnapshot()
  })

  it('has correct semantic structure', () => {
    const { container } = render(<Default />)
    const element = container.firstChild as HTMLElement
    expect(element.tagName).toBe('DIV')
  })

  it('forwards ref correctly', () => {
    const ref = { current: null }
    render(<Default ref={ref} />)
    expect(ref.current).toBeInstanceOf(HTMLDivElement)
  })
})
