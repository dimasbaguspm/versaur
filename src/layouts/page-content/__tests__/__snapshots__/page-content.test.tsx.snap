// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageContent > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full px-4 sm:px-6 py-6 sm:py-8"
  >
    <div
      class="space-y-6"
    >
      <div
        class="prose"
      >
        <p
          class="text-ghost text-left font-normal text-base"
        >
          This is the main content area of your page. PageContent provides consistent horizontal spacing that matches the PageHeader component, ensuring a cohesive layout across your application.
        </p>
      </div>
      <div
        class="grid gap-4 md:grid-cols-2 lg:grid-cols-3"
      >
        <div
          class="p-4 bg-neutral rounded-lg"
        >
          <h3
            class="text-ghost text-left text-lg font-semibold mb-2"
          >
            Card 1
          </h3>
          <p
            class="text-gray-500 text-left font-normal text-base leading-normal"
          >
            Some content in the first card using Versaur Text component.
          </p>
        </div>
        <div
          class="p-4 bg-neutral rounded-lg"
        >
          <h3
            class="text-ghost text-left text-lg font-semibold mb-2"
          >
            Card 2
          </h3>
          <p
            class="text-gray-500 text-left font-normal text-base leading-normal"
          >
            Some content in the second card with semantic typography.
          </p>
        </div>
        <div
          class="p-4 bg-neutral rounded-lg"
        >
          <h3
            class="text-ghost text-left text-lg font-semibold mb-2"
          >
            Card 3
          </h3>
          <p
            class="text-gray-500 text-left font-normal text-base leading-normal"
          >
            Some content in the third card demonstrating consistency.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
